add excludeAttr  scope to comments and threads

NakulManchanda opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/3961

## Description of Changes
- Introduced scopes in thread & comments to limit cols and data being pulled out of i/o
- changes required to improve:
   - network bandwidth
   - Database i/o load
   - Sequelize OOM error for un-expectedly large data
- in many places we only need ids, thread_id, comment_id, address_id and we are still pulling whole rows
- optimized backend to pull target columns, as comments & threads contain text data - which is expensive and bulky if not needed

Query performance might improve but that is secondary.

Endpoints
- [x] queryGlobalActivity.ts - backend optimize - same response as earlier. 
- [x] viewUserActivity.ts - backend optimize - same response as earlier. 
- [x] deleteChain.ts - backend only - needed just id. 
- [x] bulkOffchain.ts - backend optimize - same response as earlier. 
- [x] searchDiscussions.ts - was already excluding added few extra - same response as earlier. 
- [x] getLinks.ts - backend optimize - same response as earlier. 
- [x] getProfiles.ts - removed version history - body, text still included. 
- [x] viewSubscriptions.ts - excludedAttributes - text, version history everything. 
- [x] getThreads.ts - include_comments - readonly - just excluding - version history. 
- [x] getNewProfile.ts - excluding canvas, version history - keeping text and body

**Models**. 
**hasMany** relationships are scoped to exclude text, version history and canvas, Its likely only ids are needed when secondary. 
- [x] Models - Address, Chain, Topic and Threads for comments - scoped inclusions

## Test Plan
- CA (click around) tested on local and frack:
  - TODO page

## Deployment Plan
<!--- Omit if unneeded -->
1. Merge to master

## Other Considerations
<!--- Follow-up tickets, breaking changes, etc -->
- might get broken down into smaller tickets for some of the functionality

NakulManchanda commented:
@timolegros would be super helpful, if you can help with some click around testing on this one!
Columns are included in most places, wherever comments & threads are not the main entity.
I want to make sure anything in UI is not dependent on some internal properties which are excluded now!

timolegros reviewed (commented):
This looks good though I am wondering if using a default scope could be useful here. @NakulManchanda are there many more thread instances that you did not have to edit the scope for?

Basic thread/comment fetching/creating appears to be working correctly but I will do a more thorough review tomorrow and CA to test out each of the routes affected. My main worry is that the client may be caching/storing some of these thread results (even when returned from routes unrelated to threads) which could cause unforeseen side effects when data is missing.

This PR will benefit from reviews from people much more familiar with the client than I am.

NakulManchanda commented:
> My main worry is that the client may be caching/storing some of these thread results (even when returned from routes unrelated to threads) which could cause unforeseen side effects when data is missing.

That be fine @timolegros , as we are removing stuff **which are not even being referred on frontend.**
Even if its cached, it will get fetched again, and would result in 200 not 304, which will update client browser local storage too.

sometimes, frontend changes need to have more backward compatibility with cached data.
only backend backward compatible with frontend changes here would be fine i believe.

NakulManchanda commented:
> are there many more thread instances that you did not have to edit the scope for?
-  do you see any uncovered? would like to cover all.
- only two are left - `activeThreads`, and `bulkThreads`, again anyone familiar with frontend can help optimize those two.
- Did not touch those where Thread is primary model "From Thread" ones

NakulManchanda commented:
Changed place:
- packages/commonwealth/server/util/queryGlobalActivity.ts
- packages/commonwealth/server/routes/viewUserActivity.ts
- packages/commonwealth/server/routes/threads/getThreads.ts
- packages/commonwealth/server/routes/subscription/viewSubscriptions.ts
- packages/commonwealth/server/routes/searchDiscussions.ts
- packages/commonwealth/server/routes/profiles/getProfiles.ts
- packages/commonwealth/server/routes/linking/getLinks.ts
- packages/commonwealth/server/routes/getNewProfile.ts
- packages/commonwealth/server/routes/getNewProfile.ts
- packages/commonwealth/server/routes/deleteChain.ts
- packages/commonwealth/server/routes/bulkOffchain.ts
- packages/commonwealth/server/models/topic.ts

Entities - Address, Chain, Comment, Thread
- packages/commonwealth/server/models/address.ts
- packages/commonwealth/server/models/chain.ts
- packages/commonwealth/server/models/comment.ts
- packages/commonwealth/server/models/thread.ts

NakulManchanda commented:
**viewSubscriptions**
With Comment:
Comments in viewSubs - don't have excluded columns - text, plaintext, version_history, canvas data
Prod:
<img width="1083" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/ebd7cefb-9767-4f35-bfa9-3ea977702b0a">

Frick:
<img width="1083" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/031e2087-092f-43df-b8e0-9032d97e6686">




With Thread:
<img width="1083" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/62056c97-333e-46d1-adf1-4366b6b93e88">

NakulManchanda commented:
@ianrowan helped with some click around testing too
https://commonxyz.slack.com/archives/C050AE0URFH/p1685977836922679?thread_ts=1685976190.683549&cid=C050AE0URFH

NakulManchanda commented on packages/commonwealth/server/routes/threads/getThreads.ts:
```diff
@@ -74,7 +74,10 @@ export const getThreads = async (
   if (include_comments)
-    include.push({ model: models.Comment, required: false });
+    include.push({
+      model: models.Comment.scope('excludeAttributes'),
```
only comment title is included, this actually can break system - if not removed!
As getComments is paged - thats much more safer
- We cannot include all comments associated with thread - just their titles - as we have 100s of comments

NakulManchanda replied:
Not sure if any bots are using getThreads to fetch all comments at once! 
for react app getThreads - include_comments is always false

We have to test seprately if we like to include body etc! include_comments=true!

jnaviask replied:
The include_comments flag is a problem anyway. I guess we're lucky it didn't already break.

jnaviask replied:
Should cc @zakhap here

NakulManchanda replied:
What we can do is 

one of these, as these feed are read only

text - some query string format
Plaintext - readable good for external api
Body - markdown
Version history

NakulManchanda replied:
For now - changed it to 
- models.Comment.scope('readOnly')
- basically includes text, body, plaintext

We should deprecate `include_comments` or change meaning to just include ids and one of plaintext & text

NakulManchanda commented:
@jnaviask this is ready for review - tested on Frick by me & @ianrowan
Thanks @ianrowan

jnaviask commented on packages/commonwealth/server/routes/getNewProfile.ts:
```diff
@@ -61,7 +61,7 @@ const getNewProfile = async (
 
   const addressIds = [...new Set<number>(addresses.map((a) => a.id))];
-  const threads = await models.Thread.findAll({
+  const threads = await models.Thread.scope('excludeAttributes').findAll({
```
Changes to getNewProfile will likely break Profiles pages. Did not QA though.

NakulManchanda replied:
Again not breaking if it should on Frick

jnaviask replied:
Definitely broken
![2023-06-05-155633_1003x1018_scrot](https://github.com/hicommonwealth/commonwealth/assets/1860629/f7e0d3e9-f270-41b4-abb0-0c070eaff500)

NakulManchanda replied:
I see undefined! Actually I saw that too

NakulManchanda replied:
For read only screens, we can optimize too
we have the
- text - some query string format
- Plaintext - readable good for external api
- Body - markdown
- Version history

we should mention in acceptance criteria for new ticket to pull required columns only

NakulManchanda replied:
isThread ? body : comment.text

https://github.com/hicommonwealth/commonwealth/blob/ecac186ed8bdf2e75f2c01586c639d7a1338da67/packages/commonwealth/client/scripts/views/components/Profile/ProfileActivityRow.tsx#L124

jnaviask commented on packages/commonwealth/server/routes/profiles/getProfiles.ts:
This is an external route -- we shouldn't exclude threads/comments here or we break the external API.

NakulManchanda replied:
model: models.Comment.scope('readOnly') - **includes both text, plaintext** exclude version_history and canvas cols
      model: models.Threads.scope('readOnly') - **includes both body, plaintext** exclude version_history and canvas cols
      
 - externally we don't which cols they are relying on so including both text columns

jnaviask commented on packages/commonwealth/server/routes/searchDiscussions.ts:
This will likely break search.

NakulManchanda replied:
Its still deployed on Frick, i tried searching! Not sure why its not breaking if it should

jnaviask replied:
This is fine, actually.

jnaviask commented on packages/commonwealth/server/routes/viewUserActivity.ts:
Is this safe to exclude? Will we lose notifications bodies?

NakulManchanda replied:
Check my screenshot, seems like we are using Notification data for that screen instead of thread and comment.

it not sent to frontend even now, just ids

NakulManchanda replied:
https://github.com/hicommonwealth/commonwealth/pull/3966#issuecomment-1576984898

jnaviask commented on packages/commonwealth/server/util/queryGlobalActivity.ts:
Is this safe to exclude? Will we lose notifications bodies?

NakulManchanda replied:
Same as viewUserActivity - did not break
Response still looks same, seems like comment & thread not part of response even now - backend optimization only

jnaviask reviewed (changes requested):
Worried about regressions + a majority of the wins here are minimal, in that they're low-volume routes, so we should opt on the side of being careful not to cause regressions.

We may be able to break out the important routes into a separate PR, so that we can merge them one by one, but I'm not comfortable with the level of thoroughness here.

NakulManchanda commented:
> Worried about regressions + a majority of the wins here are minimal, in that they're low-volume routes, so we should opt on the side of being careful not to cause regressions.
> 
> We may be able to break out the important routes into a separate PR, so that we can merge them one by one, but I'm not comfortable with the level of thoroughness here.

This ticket is not for Query Performance improvement.

Focus here is not to fix slow routes but I/O, memory & network load metrics

This is follow up of OOM error
Improves things on different metrics:
- network bandwidth
- Database i/o load
- Sequelize OOM error for un-expectedly large data

There are many places where we just need ids, but we are still pulling all the columns - 
Eg. deleteChain, viewSubscription, viewActivities

Whatever we think is not safe I can rollback, create new grab n go tickets if ok.


Thread include Comments has to be fixed right away. As included comments as array doesnt not get paged.

NakulManchanda commented:
Ok will be rolling back 
external & getProfile

let me know if any other, thanks!

NakulManchanda commented:
- addressed comments and pushed to frick for another round of testing
- introduced new readOnly scope - includes text, body, plaintext - especially for external apis
- explicitly listing columns for activity, links, deleteChain query

NakulManchanda commented:
This ticket which is about to get merged includes integration tests for getThreads, getProfiles
https://github.com/hicommonwealth/commonwealth/pull/3391

jnaviask commented:
Changes not necessary -- ticked closed as wont fix.

NakulManchanda commented:
This ticket clearly had a goal to clean up usage for Threads & Comments to protect from potential OOM errors.

There are clearly few examples where we just need `id` we are still pulling all the data.
- imagine query for big community like osmosis - where we have over 500 or 1000 threads
- What will be more efficient and faster:

500 or 1000 whole threads data in memory
```
const threads = await models.Thread.findAll({
        where: { chain: chain.id },
      });
      
  SELECT * FROM Threads where chain = chain_id
```
or

500 or 1000 ids
```
const threads = await models.Thread.findAll({
        where: { chain: chain.id },
        attributes: ['id'],
      });
      
SELECT id FROM Threads where chain = chain_id
```

<img width="1470" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/562ac4f0-c60b-49e1-b495-9f0c918fce16">

This is to protect from future downsides.

NakulManchanda commented:
- We can always revert the risky parts and keep the parts we think are ok
- or whoever is working on updating this endpoints for any frontend or backend ticket should include these changes in there PR, so they can get in one by one!

- Additionally
  - for backend changes like this we will have more confidence in future 
  - as our test coverage is getting better everyday!
